### PR TITLE
[danfossairunit] Fixed null warnings and errors

### PR DESCRIPTION
**Pull request by kaikreuzer**
Signed-off-by: Kai Kreuzer <kai@openhab.org>

### CODE REVIEW

File: bundles/org.openhab.binding.danfossairunit/src/main/java/org/openhab/binding/danfossairunit/internal/DanfossAirUnitHandler.java [hunk 1]
```diff
@@ -86,56 +87,55 @@ public void initialize() {
...
     }
 
     private void updateAllChannels() {
-        if (hrv == null) {
-            return;
-        }
-        DanfossAirUnit danfossAirUnit = hrv;
-        logger.debug("Updating DanfossHRV data '{}'", getThing().getUID());
+        if (hrv != null) {
+            DanfossAirUnit danfossAirUnit = hrv;
```
Comment (cpmeister) [inline bundles/org.openhab.binding.danfossairunit/src/main/java/org/openhab/binding/danfossairunit/internal/DanfossAirUnitHandler.java:112]: I don't think this really fixes the null warning.
You would need to do your null check with a local variable instead of the field.
Also to reduce indentation changes I prefer the way it was written earlier, if only corrected slightly. 
```suggestion
        DanfossAirUnit danfossAirUnit = hrv;
        if (hrv == null) {
            return;
        }
```

File: bundles/org.openhab.binding.danfossairunit/src/main/java/org/openhab/binding/danfossairunit/internal/DanfossAirUnitHandler.java [hunk 2]
```diff
@@ -62,15 +60,18 @@ public void handleCommand(ChannelUID channelUID, Command command) {
             updateAllChannels();
         } else {
             try {
-                if (hrv == null) {
+                if (hrv != null) {
+                    DanfossAirUnit danfossAirUnit = hrv;
```
Comment (cpmeister) [inline bundles/org.openhab.binding.danfossairunit/src/main/java/org/openhab/binding/danfossairunit/internal/DanfossAirUnitHandler.java:64]:
```suggestion
                DanfossAirUnit danfossAirUnit = hrv;
                if (danfossAirUnit != null) {
                    
```

File: bundles/org.openhab.binding.danfossairunit/src/main/java/org/openhab/binding/danfossairunit/internal/DanfossAirUnitHandler.java [hunk 3]
```diff
@@ -86,56 +87,55 @@ public void initialize() {
         valueCache = new ValueCache(config.updateUnchangedValuesEveryMillis);
         try {
             hrv = new DanfossAirUnit(InetAddress.getByName(config.host), 30046);
+            DanfossAirUnit danfossAirUnit = hrv;
```
Comment (cpmeister) [inline bundles/org.openhab.binding.danfossairunit/src/main/java/org/openhab/binding/danfossairunit/internal/DanfossAirUnitHandler.java:90]: By assigning the non-null value the local variable first it will make the null checker happy.
```suggestion
            DanfossAirUnit danfossAirUnit = new DanfossAirUnit(InetAddress.getByName(config.host), 30046);
            hrv = danfossAirUnit;
```

File: bundles/org.openhab.binding.danfossairunit/src/main/java/org/openhab/binding/danfossairunit/internal/DanfossAirUnitHandler.java (hunk 1 above)
Comment (kaikreuzer) [reply bundles/org.openhab.binding.danfossairunit/src/main/java/org/openhab/binding/danfossairunit/internal/DanfossAirUnitHandler.java:112]: Try it and you will see that it fixes the null warning. If you do a not-null check and then assign the value you have checked to a local var, it will treat the local var as NonNull. It cannot treat `hrv` as nnNonNull, since it is a field that could potentially change its value to null at any time.

File: bundles/org.openhab.binding.danfossairunit/src/main/java/org/openhab/binding/danfossairunit/internal/DanfossAirUnitHandler.java (hunk 3 above)
Comment (kaikreuzer) [reply bundles/org.openhab.binding.danfossairunit/src/main/java/org/openhab/binding/danfossairunit/internal/DanfossAirUnitHandler.java:90]: I did the changes after seeing that the null checker was happy, trust me.

File: bundles/org.openhab.binding.danfossairunit/src/main/java/org/openhab/binding/danfossairunit/internal/DanfossAirUnitHandler.java (hunk 1 above)
Comment (martinvw) [reply bundles/org.openhab.binding.danfossairunit/src/main/java/org/openhab/binding/danfossairunit/internal/DanfossAirUnitHandler.java:112]: I must say that it also surprises me. Are both maven and Eclipse fine now? Technically that is not correct because a lot can change between checking and assigning given no locking or synchronization seems to be applied.

Comment (cpmeister) [reply bundles/org.openhab.binding.danfossairunit/src/main/java/org/openhab/binding/danfossairunit/internal/DanfossAirUnitHandler.java:112]: I'm confused, are you agreeing with me or not?

Comment (kaikreuzer) [reply bundles/org.openhab.binding.danfossairunit/src/main/java/org/openhab/binding/danfossairunit/internal/DanfossAirUnitHandler.java:112]: @martinvw Afaik the compiler makes sure that this is the way to make sure that you can safely access a field as nonnull after an if - it only works if it is the very next line after the if.

Comment (martinvw) [reply bundles/org.openhab.binding.danfossairunit/src/main/java/org/openhab/binding/danfossairunit/internal/DanfossAirUnitHandler.java:112]: > I'm confused, are you agreeing with me or not?

If you are talking to me, yes @cpmeister I agree with you.

Comment (martinvw) [reply bundles/org.openhab.binding.danfossairunit/src/main/java/org/openhab/binding/danfossairunit/internal/DanfossAirUnitHandler.java:112]: > @martinvw Afaik the compiler makes sure that this is the way to make sure that you can safely access a field as nonnull after an if - it only works if it is the very next line after the if.

What a feature, I can imagine that it fixes a lot of broken code, but is it part of the JDK/JRE specs or an implementation detail for specific JDK/JRE's?

And yes I tested it. Even adding a `System.out.print()` above it breaks it 😄

Comment (martinvw) [reply bundles/org.openhab.binding.danfossairunit/src/main/java/org/openhab/binding/danfossairunit/internal/DanfossAirUnitHandler.java:112]: I must say that given someone can break if by adding any statement above I would say its still best practice to do it really local so that it does not depend on this corner case :-)

Comment (martinvw) [reply bundles/org.openhab.binding.danfossairunit/src/main/java/org/openhab/binding/danfossairunit/internal/DanfossAirUnitHandler.java:112]: @kaikreuzer I did some more tests it's not true, my first tests give false positives because of context switching, but technically this should be the scenario you describe:

```java
class Scratch {
    private static String field = null;
    public static void main(String[] args) {
        for (int i = 0; i < 10_002; i++) {
            new Thread(new Test()).start();
        }
    }

    static class Test implements Runnable {
        @Override
        public void run() {
            while (true) {
                field = "";

                try {
                    Thread.sleep(1);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }

                if (field != null) {
                    System.out.print("- " + field.length());
                }

                field = null;
            }
        }
    }
}
```

Without the `Thread.sleep` it does not give any NPE's but with it does most likely because it causes more context switches.

Comment (kaikreuzer) [reply bundles/org.openhab.binding.danfossairunit/src/main/java/org/openhab/binding/danfossairunit/internal/DanfossAirUnitHandler.java:112]: The correct code would be 
```
                if (field != null) {
                    String localField = field;
                    System.out.print("- " + localField.length());
                }
```
Can you try this?

Comment (martinvw) [reply bundles/org.openhab.binding.danfossairunit/src/main/java/org/openhab/binding/danfossairunit/internal/DanfossAirUnitHandler.java]: Updated code:

```java
class Scratch {
    private static String field = null;
    public static void main(String[] args) {
        for (int i = 0; i < 10_002; i++) {
            new Thread(new Test()).start();
        }
    }

    static class Test implements Runnable {
        @Override
        public void run() {
            while (true) {
                field = "";

                try {
                    Thread.sleep(1);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }

                if (field != null) {
                    var local = field;
                    System.out.print("- " + local.length());
                }

                field = null;
            }
        }
    }
}
```

Output still contains NPE's

Comment (cpmeister) [reply bundles/org.openhab.binding.danfossairunit/src/main/java/org/openhab/binding/danfossairunit/internal/DanfossAirUnitHandler.java]: oof, I just realized my original suggestion was bogus due to my oversight. whoops. Sorry for the confusion.
